### ISSUE TITLE
SVN#15047 (Bugzilla Bug 2876)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#8558]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2876
Date: 2020-08-06T02:33:35+00:00
From: Seven D &lt;<Seven.ding>&gt;
To:   Maciej Rabeda &lt;<maciej.rabeda>&gt;
CC:   @jiaxinwu, @lersek, maciej.rabeda, Seven.ding, siyuan.fu

Last updated: 2021-05-07T02:55:35+00:00

**Comment by tianocore-issues** [comment]
## Comment 13326
Date: 2020-08-06 02:33:35 +0000
From: Seven D &lt;<Seven.ding>&gt;

- Industry Specification: ---
- Release Observed: UDK 2018
- Releases to Fix: UDK 2018
- Target OS: Windows
- Bugzilla Assignee(s): Maciej Rabeda &lt;<maciej.rabeda>&gt;

We found one issue related to EDK2 SVN#15047, may I know if we can revert this changed ? thanks 
EDK2 link : https://svn.code.sf.net/p/edk2/code/trunk/edk2
 
 
Issue : Cannot accept the TFTP server IP ,  but always  accept the ProxyDHCP server IP
 
 
 
How to reproduce
follow below steps to quickly see the issue:
What you need is a any 3 machines that will serve as servers, any network switch, at least 4 ports, some network cables.
 
First of all, set IP addresses as follows. To prevent any side issues, turn off firewall and try pinging the machines when connected to the switch.
 
TFTP
IP 192.168.20.7

 
DHCP
192.168.20.100

 
ProxyDHCP
192.168.20.11

**Comment by tianocore-issues** [comment]
## Comment 13330
Date: 2020-08-06 07:06:18 +0000
From: @lersek

Based on the SVN revision (#15047), this seems to be about git commit ecec42044078 (&quot;Update PXE driver to support PXE forced mode.&quot;, 2014-01-06).

**Comment by tianocore-issues** [comment]
## Comment 13339
Date: 2020-08-07 06:44:59 +0000
From: Seven D &lt;<Seven.ding>&gt;

who can help fix the issue ? @@.
i have modified some codes as following 

  if (Private-&gt;ServerIp.Addr[0] == 0 || !IS_ENABLE_USE_SERVER_LIST (VendorOpt-&gt;DiscoverCtrl)) {  
    //
    // ServerIp.Addr[0] equals zero means we failed to get IP address from boot server list.
    // Try to use next server address field.
    //
    CopyMem (
      &amp;Private-&gt;ServerIp,
      &amp;Cache4-&gt;Packet.Offer.Dhcp4.Header.ServerAddr,
      sizeof (EFI_IPv4_ADDRESS)
      );
  }

**Comment by tianocore-issues** [comment]
## Comment 13417
Date: 2020-08-17 14:02:40 +0000
From: Maciej Rabeda &lt;<maciej.rabeda>&gt;

Hello,

From the very short issue description, I assume that you are referring to PXE boot scenario with proxyDHCP. However, I do not understand the issue 100%, especially the part: &quot;cannot accept TFTP server IP&quot;.

In PXE boot scenario with proxyDHCP, we have 3 machines:
- DHCP server (gives IP and subnet)
- proxyDHCP server (gives PXE server IP and PXE-related DHCP options)
- PXE client

Which of the machines you have mentioned fit in the names above?

**Comment by tianocore-issues** [comment]
## Comment 13424
Date: 2020-08-18 06:36:20 +0000
From: Seven D &lt;<Seven.ding>&gt;

1. Prepare three machines for server and a switch with at least 4 LAN ports (you can also use a router instead). The machine will use Windows 10 system
2. Connect all three machines to the switch with a network cable. The switch does not need to be connected to an external network.
3. Turn off the firewalls on each machine
4. Set the static Address of the first machine to 192.168.20.2, extract the TFTP.rAR package and copy it to the root directory of Disk C, click Serva32. exe and wait for 7 seconds, then click &quot;Thanks, not today...&quot; The button
5. Set the second machine static IP 192.168.20.100, extract dhCp.rar and copy it to the root directory of Disk C, and click Serva32.exe
6. Set the third machine static IP 192.168.20.10, extract copy of proxydhcp.rar and proxydhcpworking. Rar to the root directory of Disk C, just click serva32.exe in the proxydhCP folder
7. Ping the machines to make sure they are connected
8. To prepare a fourth machine, connected to the switch.
9. Set UEFI only Mode in the BIOS (the fourth machine ) , Restart the machine press F12 to enter the boot page and select PXE Boot.
you will see the fourth machine cannot download file from TFTP -&gt; problem



even if adding  below W/A, but still failed .it can get pass result when revert #15047
There is an option PXE_BOOT_SERVERS(Vendor specific option 43, Tag 6, bit 2) in DHCP to use the Server list(PXE_BOOT_SERVERS.png). Usually this option bit is enabled by default, but in this case, if this bit is enabled, incorrect server(ProxyDHCP) IP is stored in the BIOS and caused to fail boot image retrieval. 
So for this work around, if this bit is disabled, BIOS does not use Server list and accept the TFTP server IP offered by ProxyDHCP and system boots up using correct boot image from TFTP server. 
 
Original: 
DHCP_Option1=43|0x06010708070a0b01C0A8140A09090a0b064555524F44420a09005058455F424F4F54ff 
Work Around: 
DHCP_Option1=43|0x06010308070a0b01C0A8140A09090a0b064555524F44420a09005058455F424F4F54ff

**Comment by tianocore-issues** [comment]
## Comment 13426
Date: 2020-08-18 07:58:40 +0000
From: @lersek

Maciej,

looking at git commit ecec42044078:

- before the patch, &quot;Private-&gt;ServerIp&quot; would be populated with
&quot;Cache4-&gt;Packet.Offer.Dhcp4.Header.ServerAddr&quot;.

- after the patch, &quot;Private-&gt;ServerIp&quot; is populated with
&quot;Cache4-&gt;Packet.Offer.Dhcp4.Header.ServerAddr&quot; only as a fallback. In
the cases when the fallback is inactive, &quot;Private-&gt;ServerIp&quot; is filled
from VendorOpt-&gt;BootSvr-&gt;IpAddr[0].

Now, per PXEv2.1 spec, PXE_DISCOVERY_CONTROL (= option 43 tag 6) bit#2
means,

  If set, *only* use/accept servers in PXE_BOOT_SERVERS. [emphasis mine]

and

  If PXE_DISCOVERY_CONTROL bit 2 is set, the client may still use
  multicast and broadcast discovery (if it is permitted by bits 0 and
  1); but the client may only accept replies from servers that are
  identified in the PXE_BOOT_SERVERS option.

and PXE_BOOT_SERVERS is explained elsewhere as &quot;Option #43 tag #8&quot;.


Commit ecec42044078 checks the following:

- IS_DISABLE_PROMPT_MENU (VendorOpt-&gt;DiscoverCtrl)
  --&gt; whether bit#3 in option#43 tag#6

- IS_VALID_BOOT_SERVERS (VendorOpt-&gt;BitMap)
  --&gt; whether option#43 tag#8 (=PXE_BOOT_SERVERS) exists

However, the commit doesn&apos;t check bit#2 in option#43 tag#6. That would
be:

  IS_ENABLE_USE_SERVER_LIST (VendorOpt-&gt;DiscoverCtrl)

So I guess the ask is for something like:

&gt;| diff --git a/NetworkPkg/UefiPxeBcDxe/PxeBcBoot.c b/NetworkPkg/UefiPxeBcDxe/PxeBcBoot.c
&gt;| index d062a526077b..75455a6b1169 100644
&gt;| --- a/NetworkPkg/UefiPxeBcDxe/PxeBcBoot.c
&gt;| +++ b/NetworkPkg/UefiPxeBcDxe/PxeBcBoot.c
&gt;| @@ -504,7 +504,9 @@ PxeBcDhcp4BootInfo (
&gt;|    // If all these fields are not available, use option 54 instead.
&gt;|    //
&gt;|    VendorOpt = &amp;Cache4-&gt;VendorOpt;
&gt;| -  if (IS_DISABLE_PROMPT_MENU (VendorOpt-&gt;DiscoverCtrl) &amp;&amp; IS_VALID_BOOT_SERVERS (VendorOpt-&gt;BitMap)) {
&gt;| +  if (IS_DISABLE_PROMPT_MENU (VendorOpt-&gt;DiscoverCtrl) &amp;&amp;
&gt;| +      IS_ENABLE_USE_SERVER_LIST (VendorOpt-&gt;DiscoverCtrl) &amp;&amp;
&gt;| +      IS_VALID_BOOT_SERVERS (VendorOpt-&gt;BitMap)) {
&gt;|      Entry = VendorOpt-&gt;BootSvr;
&gt;|      if (VendorOpt-&gt;BootSvrLen &gt;= sizeof (PXEBC_BOOT_SVR_ENTRY) &amp;&amp; Entry-&gt;IpCnt &gt; 0) {
&gt;|        CopyMem (

(note that the code in comment 2 works similarly).

The client is required to consume PXE_BOOT_SERVERS when bit#2 in option#43 tag#6 is set, but is the inverse true as well?

I.e., must the client *ignore* PXE_BOOT_SERVERS, when bit#2 in option#43 tag#6 is clear?

I&apos;m not convinced. I think the DHCP configuration is wrong. The configuration should not place the ProxyDHCP IP address into PXE_BOOT_SERVERS, regardless of bit#2 in option#43 tag#6. Either place the TFTP server address in PXE_BOOT_SERVERS, or remove PXE_BOOT_SERVERS (Option #43 tag #8) altogether.

**Comment by tianocore-issues** [comment]
## Comment 13430
Date: 2020-08-18 13:50:50 +0000
From: Maciej Rabeda &lt;<maciej.rabeda>&gt;

Seven D,

Thanks for the description, I now know what is going on.

Laszlo,

Thanks for looking deeper into this, much appreciated.

To me, by saying &quot;only use/accept servers in PXE_BOOT_SERVERS&quot;, PXE spec clearly means that it is made as a limitation of functionality (enabled -&gt; accept only PXE_BOOT_SERVERS list, disabled -&gt; accept all).

PXE_BOOT_SERVERS tag existence does not make sense when that bit is clear.

Bottom line:
1. Patch is required - in the way Laszlo made it is OK.
Basically: utilize server list when ((prompt is disabled) AND (use only PXE_BOOT_SERVERS) AND (PXE_BOOT_SERVERS tag is present)). I will submit it tomorrow.

2. If the PXE_DISCOVERY_CONTROL, bit 2 is set, PXE_BOOT_SERVERS list should contain TFTP server IP address. Is your ProxyDHCP config done in a way that TFTP server should also be running on the same machine?

**Comment by tianocore-issues** [comment]
## Comment 13431
Date: 2020-08-18 13:52:38 +0000
From: Maciej Rabeda &lt;<maciej.rabeda>&gt;

Moving to &apos;Confirmed&apos; for patching PxeBcDhcp4BootInfo().
Assigning to myself.

**Comment by tianocore-issues** [comment]
## Comment 13463
Date: 2020-08-19 12:55:12 +0000
From: Maciej Rabeda &lt;<maciej.rabeda>&gt;

Patch in review: https://edk2.groups.io/g/devel/message/64453

**Comment by tianocore-issues** [comment]
## Comment 13464
Date: 2020-08-19 15:03:39 +0000
From: @lersek

IN_PROGRESS per comment 8.

**Comment by tianocore-issues** [comment]
## Comment 13488
Date: 2020-08-21 07:21:32 +0000
From: Maciej Rabeda &lt;<maciej.rabeda>&gt;

Seven D,

After discussion and careful reconsideration, issue lays in your configuration.
If PXE_DISCOVERY_CONTROL bit 2 is set, it means that ProxyDHCP server wants to limit PXE client&apos;s discovery only to servers in PXE_BOOT_SERVERS tag.

Since PXE_BOOT_SERVERS contains IP of ProxyDHCP server instead of TFTP server, boot discovery is rightfully failing.

Try running ProxyDHCP and TFTP on the same server or modify ProxyDHCP to insert TFTP server IP into PXE_BOOT_SERVERS tag.

Rejecting.

**Comment by tianocore-issues** [comment]
## Comment 15948
Date: 2021-05-07 02:55:35 +0000
From: Seven D &lt;<Seven.ding>&gt;

hello, sorry for resume this issue. 

try with following setting, but still not work. if there has a good solution to fix it.
&quot;try running ProxyDHCP and TFTP on the same server or modify ProxyDHCP to insert TFTP server IP into PXE_BOOT_SERVERS tag.&quot;

strangely, the issue cannot be found with legacy PXE boot.
